In [164]:
import pandas as pd # type: ignore
import matplotlib.pyplot as plt  # type: ignore
import seaborn as sns  # type: ignore
from sklearn.model_selection import train_test_split  # type: ignore
from sklearn.metrics import mean_absolute_error, r2_score, root_mean_squared_error  # type: ignore
from sklearn.linear_model import LinearRegression# type: ignore
from sklearn.model_selection import GridSearchCV
import numpy as np
from sklearn.linear_model import Ridge, Lasso, ElasticNet
import warnings
warnings.filterwarnings("ignore")

In [165]:
data = pd.read_csv("../Data/processed_data.csv")
data.head()

,Price,Mileage,Brand,Model,Automatic Transmission,Air Conditioner,Power Steering,Remote Control,Years_of_usage
0,14.845130,5.707110,59,738,True,True,True,True,0
1,13.997833,11.782960,32,789,True,True,True,True,3
2,14.533351,11.736077,47,805,True,True,True,True,3
3,14.533351,11.338584,5,842,True,True,True,True,6
4,13.458837,11.225257,18,618,True,True,True,True,3


In [166]:
x = data.drop(columns="Price")
y = data["Price"]

In [167]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.30, random_state=42)

# Grid search tuning & cross validation  

In [168]:
# models = {
#     "LinearRegression": {
#         "model": LinearRegression(),
#         "param_grid": {
#             "fit_intercept": [True, False],
#         },
#     },
#     "Ridge": {
#         "model": Ridge(),
#         "param_grid": {
#             "alpha": [0.1, 1.0, 10.0],
#             "fit_intercept": [True, False],
#             "solver": ["auto", "svd", "sag", "saga"],
#             "max_iter": [100, 1000, 10000]
#         },
#     },
#     "Lasso": {
#         "model": Lasso(),
#         "param_grid": {
#             "alpha": [0.1, 1.0, 10.0],
#             "fit_intercept": [True, False],
#             "max_iter": [100, 1000, 10000]
#         },
#     },
#     "ElasticNet": {
#         "model": ElasticNet(),
#         "param_grid": {
#             "alpha": [0.1, 1.0, 10.0],
#             "l1_ratio": [0.1, 0.5, 0.9],
#             "fit_intercept": [True, False],
#             "max_iter": [100, 1000, 10000]
#         },
#     },
# }

# for model, model_info in models.items():
#     print(f"\nPerforming Grid Search for {model}...")

#     grid_search = GridSearchCV(
#         estimator=model_info["model"],
#         param_grid=model_info["param_grid"],
#         cv=5,
#         scoring="neg_mean_absolute_error",
#     )

#     grid_search.fit(x_train, y_train)

#     # Print best parameters and scores
#     print(f"Best parameters for {model}: {grid_search.best_params_}")
#     print(f"Best cross-validation score for {model}: {-grid_search.best_score_}")
#     # Evaluate the model on the test set
#     y_pred = grid_search.predict(x_test)
#     mae = mean_absolute_error(y_test, y_pred)
#     r2 = r2_score(y_test, y_pred)
#     rmse = root_mean_squared_error(y_test, y_pred)

#     print(f"Test MAE for {model}: {mae}")
#     print(f"Test R2 for {model}: {r2}")
#     print(f"Test RMSE for {model}: {rmse}")

In [169]:
model = Ridge(alpha=10.0, fit_intercept=True, solver="svd", max_iter=100)
model.fit(x_train, y_train)
y_pred = model.predict(x_test)

# Evaluation 

In [170]:
y_pred = model.predict(x_test)
mae = mean_absolute_error(y_test, y_pred) * 100
r2 = r2_score(y_test, y_pred) * 100 
rmse = root_mean_squared_error(y_test, y_pred)* 100 

print(f"Test MAE : {mae:.2f}")
print(f"Test R2 : {r2:.2f}")
print(f"Test RMSE : {rmse:.2f}")

Test MAE : 40.45
Test R2 : 61.86
Test RMSE : 53.16


In [171]:
# print the actual and predicted values with two decimal places after applying np.expm1
pd.DataFrame({'Actual': np.expm1(y_test).round(2), 'Predicted': np.expm1(y_pred).round(2)}).head(20)

,Actual,Predicted
10534,6800000.0,3071642.15
11227,250000.0,560986.11
14319,1270000.0,1041592.48
17615,820000.0,932498.59
16922,2000000.0,581209.66
14855,460000.0,400473.85
3643,4200000.0,1150722.92
1595,1450000.0,539567.73
18046,450000.0,434042.74
12528,850000.0,536719.83


# Evaluation the two models 

### Linear regression :
Test MAE : 40.45

Test R2 : 61.86

Test RMSE : 53.16

### RandomForest :

Root Mean Squared Error (RMSE): 0.30

Mean Absolute Error (MAE): 0.21

R-squared: 0.88


## It's clear that the Random forest is more better that the linear regression